core/node/fee_model/src/l1_gas_price/blob_base_fee_predictor.rs

```diff
@@ -0,0 +1,166 @@
+        .await
+        .expect("Failed to get batch stats");
+    let mut last_l1_commited_batch: Option<L1BatchNumber> = None;
+
+    for (tx_type, block) in batch_stats.mined {
+        if tx_type == AggregatedActionType::L1Batch(L1BatchAggregatedActionType::Commit) {
+            last_l1_commited_batch = Some(block.into());
+        }
+    }
+
+    let last_l1_commited_batch = last_l1_commited_batch.unwrap_or(L1BatchNumber::from(0));
+    let last_sealed_batch = connection
+        .blocks_dal()
+        .get_sealed_l1_batch_number()
+        .await
+        .expect("Failed to get last sealed batch")
+        .unwrap_or(L1BatchNumber::from(0));
+
+    let total_blobs_to_send = connection
+        .blocks_dal()
+        .get_blobs_so_far(last_sealed_batch)
+        .await
+        .expect("Failed to get blobs so far for last sealed batch")
+        - connection
+            .blocks_dal()
+            .get_blobs_so_far(last_l1_commited_batch)
+            .await
+            .expect("Failed to get blobs so far for last committed batch");
+
+    let latest_block_number = client
+        .inner
+        .block(BlockId::Number(BlockNumber::Latest))
+        .await
+        .expect("Failed to get latest block")
+        .expect("Latest block is None")
+        .number
+        .expect("Latest block number is None");
+    let last_commited_block_number = connection
+        .eth_sender_dal()
+        .get_sent_at_block_for_commited_block(last_l1_commited_batch)
+        .await
+        .expect("Failed to get sent at block for commited block")
+        .unwrap_or(latest_block_number.as_u32());
+
+    let mut total_l1_blocks_for_these_blocks = latest_block_number
+        .saturating_sub(last_commited_block_number.into())
+        .as_u64();
+    if total_l1_blocks_for_these_blocks == 0 {
+        total_l1_blocks_for_these_blocks = 1;
+    }
+
+    if total_blobs_to_send == 0 {
+        return last_known_l1_gas_price;
+    }
+
+    tracing::debug!("Predicting blob fee cap with params: blobs_total: {total_blobs_to_send}, l1_blocks_total: {total_l1_blocks_for_these_blocks}, last_known_l1_gas_price: {last_known_l1_gas_price}");
+
+    predict_blob_fee_cap(
+        total_blobs_to_send,
+        total_l1_blocks_for_these_blocks,
+        last_known_l1_gas_price,
+    )
+}
+
+// todo: revisit this
+const MIN_BASE_FEE_PER_BLOB_GAS: u64 = 1;
+const BLOB_BASE_FEE_UPDATE_FRACTION: u64 = 3_338_477;
+const BLOB_GAS_PER_BLOB: u64 = 131_072; // spec value
+const BLOB_GAS_PER_BLOCK_TARGET: u64 = 393_216; // target number of blobs per block(Ethereum uses 3)
+const CAPACITY_BLOBS_PER_BLOCK: u64 = 6; // spec max
+
+// todo: revisit this
+const SAFETY_BPS: u32 = 11000; // +10%
+const TIP: u64 = 0; // wei per blob gas
+
+fn predict_blob_fee_cap(
+    blobs_total: u64,
+    l1_blocks_total: u64,
+    last_known_l1_gas_price: u64,
+) -> u64 {
+    // --- invert median_fee -> excess_now ---
+    let mut excess = excess_from_fee(last_known_l1_gas_price);
```

Comment:
can we query it from the block info somehow? I dont like binary search here... but if JSON RPC for blocks does not expose anything like it, then okay